### PR TITLE
feat: generate hermes bytecode bundle when exporting

### PR DESCRIPTION
**Pull request by wkozyra95**
Generates Hermes bytecode when using `expo export` only if there is `hermes-engine` or `@expo/hermes-engine` package in `node_modules`.

Info about bytecode in manifest looks like this
```
  "bytecode": {
    "url": "https://test.test/bundles/android-7b8597e8ffb18adf0358382ccd2edbc9.hbc",
    "version": 62
  },
```

### CODE REVIEW

File: packages/xdl/src/Hermes.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import path from 'path';
+import fs from 'fs-extra';
+
+import spawn from '@expo/spawn-async';
+
+const binPaths: Record<string, string[]> = {
+  win32: ['win64-bin', 'hermes.exe'],
```
Comment (dsokal) [inline packages/xdl/src/Hermes.ts]:
```suggestion
  win32: 'win64-bin/hermes.exe',
```
and so on

File: packages/xdl/src/Hermes.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  const binPath = binPaths[process.platform];
+  if (!binPath) {
+    throw new Error(`Unsupported platform ${process.platform}`);
+  }
+
+  const expoHermesPath = path.join(
+    projectRoot,
+    'node_modules',
+    '@expo',
+    'hermes-engine',
```
Comment (dsokal) [inline packages/xdl/src/Hermes.ts]: `'node_modules/@expo/hermes-engine'` instead of these three lines

File: packages/xdl/src/Hermes.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    return;
+  }
+  await spawn(binPath, ['-emit-binary', '-out', outputPath, jsBundlePath, ...hermesFlags], {
+    stdio: 'pipe',
+  });
+}
+
+export async function getBytecodeVersion(projectRoot: string) {
+  const binPath = await getBinForPlatform(projectRoot);
+  if (binPath) {
```
Comment (dsokal) [inline packages/xdl/src/Hermes.ts]:
```suggestion
  if (!binPath) {
   throw new Error('...');
  }
  ...
```

File: packages/xdl/src/Hermes.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+export async function tryGenerateBytecode(
+  projectRoot: string,
+  jsBundlePath: string,
+  outputPath: string,
+  flags: string[] = []
+) {
+  const hermesFlags = ['-O', ...flags];
+  const binPath = await getBinForPlatform(projectRoot);
+  if (!binPath) {
```
Comment (dsokal) [inline packages/xdl/src/Hermes.ts]: Maybe add some warning here?

File: packages/xdl/src/Project.ts [hunk 5]
```diff
@@ -534,17 +535,24 @@ export async function exportForAppHosting(
     .update(iosBundle)
     .digest('hex');
   const iosBundleUrl = `ios-${iosBundleHash}.js`;
+  const iosBundleBytecodeUrl = `ios-${iosBundleHash}.hbc`;
```
Comment (dsokal) [inline packages/xdl/src/Project.ts]: Please remove the lines for iOS from this PR. It doesn't make sense to have them here at this point.

File: packages/xdl/src/Hermes.ts (hunk 4 above)
Comment (wkozyra95) [reply packages/xdl/src/Hermes.ts]: Hermes is not officially supported yet so, for now, it should skip quietly if unavailable

File: packages/xdl/src/Hermes.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  }
+  await spawn(binPath, ['-emit-binary', '-out', outputPath, jsBundlePath, ...hermesFlags], {
+    stdio: 'pipe',
+  });
+}
+
+export async function getBytecodeVersion(projectRoot: string) {
+  const binPath = await getBinForPlatform(projectRoot);
+  if (!binPath) {
+    throw new Error('Hermes binnary not found');
```
Comment (dsokal) [inline packages/xdl/src/Hermes.ts]:
```suggestion
    throw new Error('Hermes binary not found');
```

**Pull request closed without merging**